pkg/didcomm/common/service/did_comm_msg.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package service
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
+)
+
+const (
+	jsonID       = "@id"
+	jsonType     = "@type"
+	jsonPurpose  = "~purpose"
+	jsonThread   = "~thread"
+	jsonThreadID = "thid"
+)
+
+// Header helper structure which keeps reusable fields
+// Deprecated: Use DIDCommMsg instead of Header
+type Header struct {
+	ID     string           `json:"@id"`
+	Thread decorator.Thread `json:"~thread"`
+	Type   string           `json:"@type"`
+	// TODO revisit ~purpose, should be generic map [Issue #1037]
+	Purpose []string `json:"~purpose"`
+}
+
+// DIDCommMsg did comm msg
+type DIDCommMsg map[string]interface{}
+
+// NewDIDCommMsg returns DIDCommMsg with Header
+func NewDIDCommMsg(payload []byte) (DIDCommMsg, error) {
+	var msg DIDCommMsg
+
+	err := json.Unmarshal(payload, &msg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid payload data format: %w", err)
+	}
+
+	return msg, nil
+}
+
+// ThreadID returns msg ~thread.thid if there is no ~thread.thid returns msg @id
+// message is invalid if ~thread.thid exist and @id is absent
+func (m DIDCommMsg) ThreadID() (string, error) {
+	if m == nil {
+		return "", ErrInvalidMessage
+	}
+
+	msgID := m.ID()
+	thread, ok := m[jsonThread].(map[string]interface{})
+
+	if ok && thread[jsonThreadID] != nil {
+		var thID string
+		if v, ok := thread[jsonThreadID].(string); ok {
+			thID = v
+		}
+
+		// if message has ~thread.thid but @id is absent this is invalid message
+		if len(thID) > 0 && msgID == "" {
+			return "", ErrInvalidMessage
+		}
+
+		if len(thID) > 0 {
+			return thID, nil
+		}
+	}
+
+	// we need to return it only if there is no ~thread.thid
+	if len(msgID) > 0 {
+		return msgID, nil
+	}
+
+	return "", ErrThreadIDNotFound
+}
+
+// Purpose returns list of elements
+func (m DIDCommMsg) Purpose() []string {
```

Comment:
Please remove this method. `Purpose` will be removed from the header (see #1040 and #1058).

Reply:
@sudeshrshetty thoughts?

Reply:
@llorllale We are just removing it from header, may be replace this from key based decorator lookup?

Reply:
@sudeshrshetty by "key based decorator lookup" do you mean leave it in the map? Sure...